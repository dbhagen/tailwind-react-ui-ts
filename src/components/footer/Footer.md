Simple wrapper around the `<footer />` element

### Example

```jsx
<Footer>
  <Text is="p">&copy; E Corp {new Date().getFullYear()}</Text>
  <Text size="sm">
    Cookies are used by our fictional third party partners to track your
    browsing habits, DNA and thoughts. View our{' '}
    <Text is="a" href="#footer" text="white" text-hover="purple-dark">
      privacy policy
    </Text>{' '}
    to attempt to manage your preferences.
  </Text>
</Footer>
```
