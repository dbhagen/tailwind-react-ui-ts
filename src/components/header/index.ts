export { default as Header } from './Header'
export { default as NavBrand } from './NavBrand'
export { default as NavItem } from './NavItem'
export { default as NavMenu } from './NavMenu'
export { default as NavToggle } from './NavToggle'
