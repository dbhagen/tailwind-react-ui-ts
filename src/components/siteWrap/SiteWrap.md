By default will fill the full screen size regardless of child content lenght to create a sticky footer. Also applies baseline body text styling.

### Example

```jsx
<SiteWrap minH={0} h={64}>
  <p>Some short page content</p>
  <Footer>Footer pinned to the bottom of the screen</Footer>
</SiteWrap>
```
