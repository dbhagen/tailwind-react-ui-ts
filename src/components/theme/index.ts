export { default as defaultTheme } from './defaultTheme'
export { default as TailwindTheme } from './TailwindTheme'
export { default as TailwindThemeProvider } from './TailwindThemeProvider'
export { default as withTheme } from './withTheme'
export { default as useThemeValue } from './useThemeValue'
