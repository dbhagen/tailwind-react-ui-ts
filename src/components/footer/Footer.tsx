import React from 'react'
import PropTypes from 'prop-types'

import { Box } from '../primitives'
import { withTheme } from '../theme'
import { Container } from '../container'

const Footer = ({ theme, is, children, ...rest }) => (
  <Box
    is={is}
    role="contentinfo"
    p={{ t: theme.spacing.lg, b: theme.spacing.xl }}
    bg={theme.brandColors.secondary}
    text={theme.textColors.on.secondary}
    {...rest}
  >
    <Container padding>{children}</Container>
  </Box>
)

Footer.propTypes = {
  theme: PropTypes.shape({}).isRequired,
  is: PropTypes.oneOfType([PropTypes.string, PropTypes.func, PropTypes.object]),
  children: PropTypes.node,
}

Footer.defaultProps = {
  is: 'footer',
  children: undefined,
}

export { Footer as component }
export default withTheme(Footer)
