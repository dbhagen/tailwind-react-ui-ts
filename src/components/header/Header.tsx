import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'

import { withTheme } from '../theme'
import { Box, Flex } from '../primitives'
import { getUniqueID, filterProps } from '../utils'
import { Container } from '../container'

class Header extends PureComponent {
  constructor(props) {
    const { id, screen } = props

    super(props)

    this.state = {
      open: false,
      collapsable: !!screen,
    }

    this.mql = null
    this.id = id || getUniqueID('header')

    this.handleMatch = this.handleMatch.bind(this)
    this.handleToggle = this.handleToggle.bind(this)
  }

  componentDidMount() {
    const { theme, screen } = this.props

    if (screen && window.matchMedia) {
      this.mql = window.matchMedia(`(min-width: ${theme.breakpoints[screen]})`)
      this.mql.addListener(this.handleMatch)

      if (this.mql.matches) {
        this.handleMatch(this.mql)
      }
    }
  }

  componentWillUnmount() {
    if (this.mql) this.mql.removeListener(this.handleMatch)
  }

  handleToggle(forceState) {
    const { open } = this.state

    this.setState({
      open: forceState || !open,
    })
  }

  handleMatch(mql) {
    this.setState({ collapsable: !mql.matches })
  }

  render() {
    const { open, collapsable } = this.state
    const { theme, is, children, bg, text, screen, ...rest } = this.props

    const headerProps = {
      id: this.id,
      style: {
        bg: bg || theme.brandColors.primary,
        text: text || theme.textColors.on.primary,
      },
      open,
      collapsable,
      onToggle: this.handleToggle,
      screen,
    }

    return (
      <Box
        is={is}
        id={headerProps.id}
        bg={headerProps.style.bg}
        text={headerProps.style.text}
        p={{ y: theme.spacing.md }}
        role="banner"
        {...filterProps(rest, ['id'])}
      >
        <Container is={Flex} wrap items="center" justify="between" padding>
          {React.Children.map(children, child =>
            React.cloneElement(child, { header: headerProps }),
          )}
        </Container>
      </Box>
    )
  }
}

Header.propTypes = {
  theme: PropTypes.shape({}).isRequired,
  is: PropTypes.oneOfType([PropTypes.string, PropTypes.func, PropTypes.object]),
  children: PropTypes.node,
  bg: PropTypes.oneOfType([PropTypes.string, PropTypes.array]),
  text: PropTypes.oneOfType([PropTypes.string, PropTypes.array]),
  id: PropTypes.string,
  screen: PropTypes.oneOfType([PropTypes.bool, PropTypes.string]),
}

Header.defaultProps = {
  is: 'header',
  children: undefined,
  bg: undefined,
  text: undefined,
  id: undefined,
  screen: 'lg',
}

export { Header as component }
export default withTheme(Header)
