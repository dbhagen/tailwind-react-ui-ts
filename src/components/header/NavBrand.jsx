import React from 'react'
import PropTypes from 'prop-types'

import { withTheme } from '../theme'
import { Box } from '../primitives'

const NavBrand = ({
  theme,
  header: { style, screen },
  is,
  children,
  ...rest
}) => {
  const responsive = screen
    ? {
        [`m-${screen}`]: { r: theme.spacing.lg },
      }
    : {}

  const aria = !(typeof is === 'string' && is.startsWith('h'))
    ? {
        role: 'heading',
        'aria-level': 1,
      }
    : {}

  return (
    <Box
      is={is}
      inlineBlock
      flex={[true, 'no-shrink']}
      items="center"
      h={12}
      text={style.text || theme.textColors.on.primary}
      {...responsive}
      {...aria}
      {...rest}
    >
      {children}
    </Box>
  )
}

NavBrand.propTypes = {
  theme: PropTypes.shape({}).isRequired,
  is: PropTypes.oneOfType([PropTypes.string, PropTypes.func, PropTypes.object]),
  children: PropTypes.node,
  header: PropTypes.shape({
    style: PropTypes.object,
    screen: PropTypes.oneOfType([PropTypes.bool, PropTypes.string]),
  }),
}

NavBrand.defaultProps = {
  is: 'div',
  children: undefined,
  header: { style: {}, screen: 'lg' },
}

export { NavBrand as component }
export default withTheme(NavBrand)
