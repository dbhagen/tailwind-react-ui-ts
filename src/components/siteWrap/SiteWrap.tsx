import React from 'react'
import PropTypes from 'prop-types'

import { Flex, Box } from '../primitives'
import Footer, { component as FooterComponent } from '../footer/Footer'
import { withTheme } from '../theme'

const SiteWrap = ({ is, theme, children, ...rest }) => {
  let footer

  React.Children.forEach(children, child => {
    const { type } = child
    if (
      type &&
      (type === Footer || type.displayName === FooterComponent.displayName)
    ) {
      footer = child
    }
  })

  return (
    <Flex
      is={is}
      col
      minH="screen"
      leading="normal"
      font={theme.text.family.body}
      text={[theme.text.size.body[1], theme.textColors.body]}
      {...rest}
    >
      <Box flex={['auto', 'no-shrink']}>
        {React.Children.map(children, child => {
          if (child === footer) return false
          return child
        })}
      </Box>
      <Box flex={['auto', 'no-shrink', 'no-grow']} m={{ t: theme.spacing.lg }}>
        {footer}
      </Box>
    </Flex>
  )
}

SiteWrap.propTypes = {
  is: PropTypes.oneOfType([PropTypes.string, PropTypes.func, PropTypes.object]),
  theme: PropTypes.shape({}).isRequired,
  children: PropTypes.node,
}

SiteWrap.defaultProps = {
  is: 'div',
  children: undefined,
}

export { SiteWrap as component }
export default withTheme(SiteWrap)
