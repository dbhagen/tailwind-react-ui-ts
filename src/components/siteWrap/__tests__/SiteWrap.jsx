import React from 'react'
import { shallow } from 'enzyme'

import { defaultTheme } from '../../theme'
import { component as SiteWrap } from '../SiteWrap'
import { Header } from '../../header'
import { Footer } from '../../footer'

const setup = (testProps = {}) => {
  const props = Object.assign({ theme: defaultTheme }, testProps)

  const wrapper = shallow(
    <SiteWrap {...props}>
      <Header>Header</Header>
      <p>Body</p>
      <Footer>Footer</Footer>
    </SiteWrap>,
  )

  return {
    props,
    wrapper,
  }
}

describe('SiteWrap', () => {
  it('renders matching snapshot', () => {
    const { wrapper } = setup()

    expect(wrapper).toMatchSnapshot()
  })
})
