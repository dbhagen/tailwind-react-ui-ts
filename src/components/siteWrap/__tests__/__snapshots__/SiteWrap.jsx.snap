// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SiteWrap renders matching snapshot 1`] = `
<Flex
  col={true}
  font="sans"
  inline={false}
  inlineFlex={false}
  is="div"
  leading="normal"
  minH="screen"
  reverse={false}
  text={
    Array [
      "base",
      "gray-900",
    ]
  }
  wrap={false}
  wrapReverse={false}
>
  <Box
    flex={
      Array [
        "auto",
        "no-shrink",
      ]
    }
    inline={false}
    inlineBlock={false}
    is="div"
  >
    <WithTheme(Header)
      key=".0"
    >
      Header
    </WithTheme(Header)>
    <p
      key=".1"
    >
      Body
    </p>
  </Box>
  <Box
    flex={
      Array [
        "auto",
        "no-shrink",
        "no-grow",
      ]
    }
    inline={false}
    inlineBlock={false}
    is="div"
    m={
      Object {
        "t": 8,
      }
    }
  >
    <WithTheme(Footer)>
      Footer
    </WithTheme(Footer)>
  </Box>
</Flex>
`;
